### ISSUE TITLE
Can I use adam as optimizer 

### ISSUE THREAD

**Issue by mytxgmy** [facebookresearch/Detectron#676]
I'm trying to using Adam instead of SGD,I have changed add_single_gpu_param_update_ops in optimizer.py and broadcast_parameters in net.py.But I get an error:
`E0922 05:16:10.371023  1615 net_async_base.cc:376] [enforce fail at adam_op.h:91] OperatorBase::InputIsTensorType(ITER, CPU). Error from operator: 
input: "gpu_0/rpn_cls_logits_w" input: "gpu_0/rpn_cls_logits_w_first_moment" input: "gpu_0/rpn_cls_logits_w_second_moment" input: "gpu_0/rpn_cls_logits_w_grad" input: "gpu_0/lr" input: "g
pu_0/optimizer_iteration" output: "gpu_0/rpn_cls_logits_w" output: "gpu_0/rpn_cls_logits_w_first_moment" output: "gpu_0/rpn_cls_logits_w_second_moment" name: "" type: "Adam" device_option
 { device_type: 1 cuda_gpu_id: 0 },  op Adam
`

**Comment by ir413** [comment]
Hi @mytxgmy, the error you're seeing is due to the `iter` blob being on CPU. Namely, [`Adam`](https://caffe2.ai/docs/operators-catalogue.html#adam) op requires that its input `iter` blob is a `CPU` tensor (see [here](https://github.com/pytorch/pytorch/blob/master/caffe2/sgd/adam_op.h#L91)).

I suspect that you may be creating the `iter` blob in the [`add_single_gpu_param_update_ops`](https://github.com/facebookresearch/Detectron/blob/master/detectron/modeling/optimizer.py#L90) function analogously to how the `lr` blob is created? If so, the blob ends up being created on the GPU (see [here](https://github.com/facebookresearch/Detectron/blob/master/detectron/modeling/optimizer.py#L48)) and results in the error due to the requirement described above. I think that the easiest way to get around this may be to use the [`Iter`](https://caffe2.ai/docs/operators-catalogue.html#iter) op. For example, add something like `cur_iter = mode.Iter('cur_iter')` in the `add_single_gpu_param_update_ops` function and use `cur_iter` in the `Adam` op.

**Comment by ir413** [comment]
In case it helps, here is a patch that you can use to get the code to run with Adam. Please note that I have not tried to train a model using this and have only checked that the code runs without errors (so there may still be issues).

```
diff --git a/detectron/detectron/modeling/optimizer.py b/detectron/detectron/modeling/optimizer.py
--- a/detectron/detectron/modeling/optimizer.py
+++ b/detectron/detectron/modeling/optimizer.py
@@ -93,6 +93,9 @@
     lr = model.param_init_net.ConstantFill(
         [], 'lr', shape=[1], value=0.0
     )
+    cur_iter = model.Iter(
+        'cur_iter'
+    )
     one = model.param_init_net.ConstantFill(
         [], 'one', shape=[1], value=1.0
     )
@@ -106,9 +109,12 @@
     for param in model.TrainableParams(gpu_id=gpu_id):
         logger.debug('param ' + str(param) + ' will be updated')
         param_grad = model.param_to_grad[param]
-        # Initialize momentum vector
-        param_momentum = model.param_init_net.ConstantFill(
-            [param], param + '_momentum', value=0.0
+        # Initialize moments
+        param_moment1 = model.param_init_net.ConstantFill(
+            [param], param + '_moment1', value=0.0
+        )
+        param_moment2 = model.param_init_net.ConstantFill(
+            [param], param + '_moment2', value=0.0
         )
         if param in model.biases:
             # Special treatment for biases (mainly to match historical impl.
@@ -122,9 +128,8 @@
         elif cfg.SOLVER.WEIGHT_DECAY > 0:
             # Apply weight decay to non-bias weights
             model.WeightedSum([param_grad, one, param, wd], param_grad)
-        # Update param_grad and param_momentum in place
-        model.net.MomentumSGDUpdate(
-            [param_grad, param_momentum, lr, param],
-            [param_grad, param_momentum, param],
-            momentum=cfg.SOLVER.MOMENTUM
+        # Update param_grad and moments in place
+        model.net.Adam(
+            [param, param_moment1, param_moment2, param_grad, lr, cur_iter],
+            [param, param_moment1, param_moment2]
         )
diff --git a/detectron/detectron/utils/net.py b/detectron/detectron/utils/net.py
--- a/detectron/detectron/utils/net.py
+++ b/detectron/detectron/utils/net.py
@@ -192,7 +192,8 @@
                     workspace.FeedBlob(p, data)
 
     _do_broadcast(model.params)
-    _do_broadcast([b + '_momentum' for b in model.TrainableParams()])
+    _do_broadcast([b + '_moment1' for b in model.TrainableParams()])
+    _do_broadcast([b + '_moment2' for b in model.TrainableParams()])
 
 
 def sum_multi_gpu_blob(blob_name):

```

**Comment by mytxgmy** [comment]
@ir413 Thanks a lot .

**Comment by mytxgmy** [comment]
@ir413 But when I use adam in e2e_faster_rcnn_R-50-C4_1x or my models,the loss becomes higher and higher.
`e2e_faster_rcnn_R-50-C4_1x:
json_stats: {"accuracy_cls": 0.003174, "eta": "74 days, 12:57:54", "iter": 0, "loss": 5.250670, "loss_bbox": 0.076
905, "loss_cls": 4.409735, "loss_rpn_bbox": 0.085126, "loss_rpn_cls": 0.678904, "lr": 0.003333, "mb_qsize": 64, "m
em": 6418, "time": 35.779300}
json_stats: {"accuracy_cls": 0.000000, "eta": "4 days, 1:14:25", "iter": 20, "loss": 658416300275040807550976.0000
00, "loss_bbox": 311640278980717698875392.000000, "loss_cls": 365829543696527832121344.000000, "loss_rpn_bbox": 18
9047219431669760.000000, "loss_rpn_cls": 431318379849056256.000000, "lr": 0.003600, "mb_qsize": 48, "mem": 6418, "
time": 1.945022}
json_stats: {"accuracy_cls": 0.000000, "eta": "9:58:18", "iter": 40, "loss": 1125158371714547526402048.000000, "lo
ss_bbox": 440954086828895262212096.000000, "loss_cls": 676166635756321700315136.000000, "loss_rpn_bbox": 362087685
1817218048.000000, "loss_rpn_cls": 8053276779097358336.000000, "lr": 0.003867, "mb_qsize": 4, "mem": 6418, "time":
 0.199483}
json_stats: {"accuracy_cls": 0.000000, "eta": "9:33:04", "iter": 60, "loss": 1430689473507057251385344.000000, "lo
ss_bbox": 566101990149411311714304.000000, "loss_cls": 864148735182213052628992.000000, "loss_rpn_bbox": 150895143
30523172864.000000, "loss_rpn_cls": 32561263212717146112.000000, "lr": 0.004133, "mb_qsize": 1, "mem": 6418, "time
": 0.191086}

`

**Comment by mytxgmy** [comment]
@ir413 But when I use adam in e2e_faster_rcnn_R-50-C4_1x or my experiments,the loss becomes higher and higher.
`e2e_faster_rcnn_R-50-C4_1x:
json_stats: {"accuracy_cls": 0.003174, "eta": "74 days, 12:57:54", "iter": 0, "loss": 5.250670, "loss_bbox": 0.076
905, "loss_cls": 4.409735, "loss_rpn_bbox": 0.085126, "loss_rpn_cls": 0.678904, "lr": 0.003333, "mb_qsize": 64, "m
em": 6418, "time": 35.779300}
json_stats: {"accuracy_cls": 0.000000, "eta": "4 days, 1:14:25", "iter": 20, "loss": 658416300275040807550976.0000
00, "loss_bbox": 311640278980717698875392.000000, "loss_cls": 365829543696527832121344.000000, "loss_rpn_bbox": 18
9047219431669760.000000, "loss_rpn_cls": 431318379849056256.000000, "lr": 0.003600, "mb_qsize": 48, "mem": 6418, "
time": 1.945022}
json_stats: {"accuracy_cls": 0.000000, "eta": "9:58:18", "iter": 40, "loss": 1125158371714547526402048.000000, "lo
ss_bbox": 440954086828895262212096.000000, "loss_cls": 676166635756321700315136.000000, "loss_rpn_bbox": 362087685
1817218048.000000, "loss_rpn_cls": 8053276779097358336.000000, "lr": 0.003867, "mb_qsize": 4, "mem": 6418, "time":
 0.199483}
`

**Comment by ir413** [comment]
@mytxgmy: we do not use adam in our standard configurations so you will likely need to experiment with hyper-parameters in order to find a good setting for adam and get the model to converge.
